### DISCUSSION TITLE
How to add localStorage logic to custom components (best practice?)

### DISCUSSION THREAD

**Discussion by ZenaMel** [facebook/docusaurus#8775] [Q&A] [answered]
Should we use the default React localStorage API or does Docusaurus have utils we can and should use?

I'd like to extend the capabilities of my docs site by using localStorage. I know that Docusaurus uses localStorage under the hood in components like Tabs, but I haven't seen official docs on util functions we can use to embellish our own components with localStorage functionality. Should we use React's features or does Docusaurus have internals for this we should use instead?

**Comment by slorber** [comment, marked as answer]
Docusaurus has internal things but they are not documented on purpose.

Use at your own risk until we officially expose these APIs

```js
import {useStorageSlot} from "@docusaurus/theme-common"

const [value, api] = useStorageSlot(key);

api.set("newValue");
```



Nothing prevents you from using localStorage like in any React app, but:
- you are more likely to encounter hydration issues due to bad usage like `useState(() => localStorage.getItem("key"))`
- you won't get localStorage sync between browser tabs
- you won't get fail-safe behavior in case localStorage throws (Docusaurus embedded in iframe with some strict privacy browser settings)

That's what the Docusaurus internal helpers solve, nothing more.


See also related tweets
- https://twitter.com/sebastienlorber/status/1615329010761842689
- https://twitter.com/sebastienlorber/status/1629154621489913856

**Reply by ZenaMel** [reply]
Thanks a bunch Sebastien!
